File: apps/mcp/lib/mcp/misc/reading.ex
defmodule Mcp.Reading do
  def license, do: """
     Master Control Program for Wiss Landing
     Copyright (C) 2016  Tim Hughey (thughey)

     This program is free software: you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published by
     the Free Software Foundation, either version 3 of the License, or
     (at your option) any later version.

     This program is distributed in the hope that it will be useful,
     but WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
     GNU General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>
     """

  @moduledoc """
    This module implements Owfs as a GenServer
  """

  require Logger

  use Timex
  alias Mcp.Reading

  defstruct name: "no_name", kind: "no_kind",
    read_at: Timex.zero(), read_us: 0, val: 0.0, ttl: 0

  @name :name
  @kind :kind
  @val :val
  @read_at :read_at
  @read_us :read_us
  @ttl :ttl

  @def_ttl 30_000   # in milliseconds

  @spec kinds() :: [String.t]
  def kinds, do: ["temperature", "humidity"]

  def create(n, k, v) when is_tuple(v) do
    create(n, k, v, @def_ttl)
  end
  def create(n, k, v) when is_number(v) do
    create(n, k, {1, {:ok, v}})
  end
  def create(name, kind, {read_us, {:ok, val}}, ttl)
  when is_binary(name) and is_binary(kind) and is_float(val) and
  is_integer(read_us) and is_integer(ttl) do
    %Reading{@name => name, @kind => kind,
      @val => val, @read_at => Timex.now(),
      @read_us => read_us, @ttl => ttl}
  end
  def create(name, kind, {read_us, {:error, _val}}, ttl)
  when is_binary(name) and is_binary(kind) and
  is_integer(read_us) and is_integer(ttl) do
    %Reading{@name => name, @kind => kind,
      @val => :error, @read_at => Timex.now(),
      @read_us => read_us, @ttl => ttl}
  end

  def invalid?(%Reading{} = r), do: not valid?(r)
  def valid?(%Reading{@val => v} = r)
  when is_number(v) do
     current?(r)
  end
  def valid?(%Reading{@name => "no_name", @kind => "no_kind"}), do: false
  def valid?(%Reading{@val => :error}), do: false
  def valid?(%Reading{}, _ttl), do: false

  def current?(%Reading{} = r) do
    ttl = r.ttl

    case Timex.diff(Timex.now(), r.read_at, :milliseconds) do
      x when x < ttl  -> :true
      x when x >= ttl -> :false
    end
  end

  def name(%Reading{@name => n}), do: n
  def kind(%Reading{@kind => k}), do: k
  def val(%Reading{@val => v}), do: v
  def read_us(%Reading{@read_us => rus}), do: rus
  def read_at(%Reading{@read_at => read_at}), do: read_at
  def read_at_ns(%Reading{@read_at => read_at}) do
    (read_at |> Timex.to_unix()) * trunc(:math.pow(10,9))
  end

  def humidity?(%Reading{@kind => "humidity"}), do: true
  def humidity?(%Reading{}), do: false

  def temperature?(%Reading{@kind => "temperature"}), do: true
  def temperature?(%Reading{}), do: false

  def if_valid_execute(%Reading{@val => v} = r, func)
  when is_number(v) and is_function(func, 2) do
    func.(r, :true)
  end 
  def if_valid_execute(%Reading{@val => :error} = r, func)
  when is_function(func, 2) do
    func.(r, :false)
  end 

end

File: mcr_esp/components/mcr/src/readings/reading.cpp
/*
    reading.cpp - Readings used within Master Control Remote
    Copyright (C) 2017  Tim Hughey

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

    https://www.wisslanding.com
*/

#include <string>

#include <esp_log.h>
#include <freertos/FreeRTOS.h>
#include <sys/time.h>
#include <time.h>

#include "net/mcr_net.hpp"
#include "readings/reading.hpp"

namespace mcr {

Reading::Reading(time_t mtime) { _mtime = mtime; }

Reading::Reading(const std::string &id, time_t mtime) {
  _id = id;
  _mtime = mtime;
}

Reading::~Reading() {
  if (_json != nullptr) {
    // ESP_LOGD("Reading", "freeing _json (%p)", (void *)_json);
    delete _json;
  }
}

void Reading::commonJSON(JsonDocument &doc) {
  doc["host"] = mcr::Net::hostID();
  doc["name"] = mcr::Net::getName();
  doc["mtime"] = _mtime;
  doc["type"] = typeString(_type);

  if (_id.length() > 0) {
    doc["device"] = _id;
  }

  if (_cmd_ack) {
    doc["cmdack"] = _cmd_ack;
    doc["latency"] = _latency;
    doc["refid"] = _refid;
  }

  if (_crc_mismatches > 0) {
    doc["crc_mismatches"] = _crc_mismatches;
  }

  if (_read_errors > 0) {
    doc["read_errors"] = _read_errors;
  }

  if (_write_errors > 0) {
    doc["write_errors"] = _write_errors;
  }

  if (_read_us > 0) {
    doc["read_us"] = _read_us;
  }

  if (_write_us > 0) {
    doc["write_us"] = _write_us;
  }
}

std::string *Reading::json(char *buffer, size_t len) {
  // FUTURE IMPROVEMENT
  // std::unique_ptr<std::string> json_S(new std::string());
  // json_S.get()->reserve(1024);
  std::string *json_string = new std::string;
  json_string->reserve(2048);

  DynamicJsonDocument doc(2048);

  commonJSON(doc);
  populateJSON(doc);

  serializeJson(doc, *json_string);

  return json_string;
}

void Reading::setCmdAck(time_t latency, mcrRefID_t &refid) {
  _cmd_ack = true;
  _latency = latency;

  _refid = refid;
}

static const char *__type_string[] = {
    "base", "mcr_stat", "ph",   "stats", "remote_runtime", "relhum", "soil",
    "boot", "switch",   "temp", "text"};

const char *Reading::typeString(ReadingType_t index) {
  return __type_string[index];
}
} // namespace mcr
